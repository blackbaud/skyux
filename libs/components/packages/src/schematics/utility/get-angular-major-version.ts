import fs from 'node:fs';
import { createRequire } from 'node:module';

/**
 * Get the currently installed version of Angular.
 * Use `require` to satisfy Jest when importing from @angular/cli ESM module.
 */
export function getAngularMajorVersion(): string {
  const require = createRequire(__filename);

  try {
    const angularCliPackagePath = require.resolve('@angular/cli/package.json');
    const angularCliPackage = JSON.parse(
      fs.readFileSync(angularCliPackagePath, 'utf8'),
    );

    return angularCliPackage.version.split('.')[0];
  } catch {
    throw new Error(
      'Unable to determine Angular CLI version. Please ensure @angular/cli is installed.',
    );
  }
}
