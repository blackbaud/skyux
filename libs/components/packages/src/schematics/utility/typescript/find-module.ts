import { Tree } from '@angular-devkit/schematics';
import {
  getDecoratorMetadata,
  getMetadataField,
  isImported,
  parseSourceFile,
} from '@angular/cdk/schematics';
import ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';
import { buildRelativePath } from '@schematics/angular/utility/find-module';

import { visitProjectFiles } from '../visit-project-files';

function findModuleInPath(
  tree: Tree,
  path: string,
  filter: (sourceFile: ts.SourceFile) => boolean,
): {
  filepath: string;
  module: { className: string; metadata: ts.ObjectLiteralElement };
} | null {
  let found: {
    filepath: string;
    module: { className: string; metadata: ts.ObjectLiteralElement };
  } | null = null;
  visitProjectFiles(tree, path, (filePath) => {
    if (found) {
      return; // Stop searching if we already found a module
    }
    if (filePath.endsWith('.ts')) {
      const sourceFile = parseSourceFile(tree, filePath);
      if (
        isImported(sourceFile, 'NgModule', '@angular/core') &&
        filter(sourceFile)
      ) {
        const metadata = getDecoratorMetadata(
          sourceFile,
          'NgModule',
          '@angular/core',
        )[0] as ts.ObjectLiteralElement;
        const className =
          findClassDeclarationParent(metadata)?.name?.text || '';
        found = {
          filepath: buildRelativePath('/', `/${filePath}`),
          module: {
            className,
            metadata,
          },
        };
      }
    }
  });
  return found;
}

function findClassDeclarationParent(
  node: ts.Node | undefined,
): ts.ClassDeclaration | undefined {
  if (node && ts.isClassDeclaration(node)) {
    return node;
  }
  if (node?.parent) {
    return findClassDeclarationParent(node.parent);
  }
  return undefined;
}

export function isStandaloneComponent(
  metadata: ts.ObjectLiteralExpression,
): boolean {
  const standalone = getMetadataField(metadata, 'standalone');
  return (
    !standalone[0] ||
    (ts.isPropertyAssignment(standalone[0]) &&
      standalone[0].initializer.kind === ts.SyntaxKind.TrueKeyword)
  );
}

export function findDeclaringModule(
  tree: Tree,
  path: string,
  componentPath: string,
): {
  filepath: string;
  module: { className: string; metadata: ts.ObjectLiteralElement };
} | null {
  const source = parseSourceFile(tree, componentPath);
  const metadata = getDecoratorMetadata(
    source,
    'Component',
    '@angular/core',
  )[0] as ts.ObjectLiteralElement | undefined;
  const componentClass = findClassDeclarationParent(metadata);

  const componentClassName = componentClass?.name?.text;
  if (
    componentClassName &&
    metadata &&
    ts.isObjectLiteralExpression(metadata)
  ) {
    if (isStandaloneComponent(metadata)) {
      return {
        filepath: buildRelativePath('/', `/${componentPath}`),
        module: {
          className: componentClass.name.text,
          metadata: metadata,
        },
      };
    }

    return findModuleInPath(tree, path, (sourceFile) => {
      const moduleMetadata = getDecoratorMetadata(
        sourceFile,
        'NgModule',
        '@angular/core',
      )[0] as ts.ObjectLiteralExpression | undefined;
      if (!moduleMetadata) {
        return false;
      }
      const declarations = getMetadataField(moduleMetadata, 'declarations');
      return (
        declarations &&
        declarations.some(
          (declaration) =>
            ts.isPropertyAssignment(declaration) &&
            ts.isArrayLiteralExpression(declaration.initializer) &&
            declaration.initializer.elements.some(
              (item) => item.getText(sourceFile) === componentClassName,
            ),
        )
      );
    });
  }

  return null;
}
