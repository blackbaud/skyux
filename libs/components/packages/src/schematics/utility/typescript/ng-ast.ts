import { tags } from '@angular-devkit/core';
import {
  getDecoratorMetadata,
  getMetadataField,
  isImported,
} from '@angular/cdk/schematics';
import ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';
import { findNodes, insertImport } from '@schematics/angular/utility/ast-utils';
import { Change, InsertChange } from '@schematics/angular/utility/change';

export function getInlineTemplates(
  sourceFile: ts.SourceFile,
): { start: number; end: number }[] {
  if (isImported(sourceFile, 'Component', '@angular/core')) {
    const components = getDecoratorMetadata(
      sourceFile,
      'Component',
      '@angular/core',
    );
    const templates = components
      .filter((component) => ts.isObjectLiteralExpression(component))
      .flatMap((component) => getMetadataField(component, 'template'))
      .filter((template) => ts.isPropertyAssignment(template));
    if (templates.length > 0) {
      return templates
        .map((template) => {
          if (ts.isStringLiteralLike(template.initializer)) {
            return {
              start: template.initializer.getStart() + 1,
              end: template.initializer.getEnd() - 1, // Exclude quotes
            };
          }
          return undefined;
        })
        .filter(Boolean) as { start: number; end: number }[];
    }
  }
  return [];
}

export function getTemplateUrls(sourceFile: ts.SourceFile): string[] {
  if (isImported(sourceFile, 'Component', '@angular/core')) {
    const components = getDecoratorMetadata(
      sourceFile,
      'Component',
      '@angular/core',
    );
    const templates = components
      .filter((component) => ts.isObjectLiteralExpression(component))
      .flatMap((component) => getMetadataField(component, 'templateUrl'))
      .filter((template) => ts.isPropertyAssignment(template));
    if (templates.length > 0) {
      return templates
        .map((template) => {
          if (ts.isStringLiteralLike(template.initializer)) {
            const quoted = template.initializer.getText();
            return quoted.substring(1, quoted.length - 1);
          }
          return undefined;
        })
        .filter(Boolean) as string[];
    }
  }
  return [];
}

/**
 * Like `getDecoratorMetadata` from @angular/cdk/schematics, but for `TestBed.configureTestingModule`.
 */
export function getTestingModuleMetadata(
  source: ts.SourceFile,
): ts.ObjectLiteralExpression[] {
  return findNodes(
    source,
    (node: ts.Node): node is ts.ObjectLiteralExpression =>
      ts.isObjectLiteralExpression(node) &&
      ts.isCallExpression(node.parent) &&
      node.parent.expression.getText().trim() ===
        'TestBed.configureTestingModule',
  );
}

/**
 * Determines whether a component is standalone from its metadata.
 */
export function isStandaloneComponent(
  metadata: ts.ObjectLiteralExpression,
): boolean {
  const standalone = getMetadataField(metadata, 'standalone');
  return (
    !standalone[0] ||
    (ts.isPropertyAssignment(standalone[0]) &&
      standalone[0].initializer.kind === ts.SyntaxKind.TrueKeyword)
  );
}

/**
 * Adds a symbol to the metadata of an Angular class (NgModule, Component, Directive).
 *
 * Derived from https://github.com/angular/angular-cli/blob/20.1.x/packages/schematics/angular/utility/ast-utils.ts
 */
export function addSymbolToClassMetadata(
  source: ts.SourceFile,
  decorator:
    | 'NgModule'
    | 'Component'
    | 'Directive'
    | 'TestBed.configureTestingModule',
  filePath: string,
  metadataField: string,
  symbolName: string,
  importPath: string | null = null,
): Change[] {
  const nodes =
    'TestBed.configureTestingModule' === decorator
      ? getTestingModuleMetadata(source)
      : getDecoratorMetadata(source, decorator, '@angular/core');
  let insertedImport = false;
  return nodes
    .filter((node) => !!node && ts.isObjectLiteralExpression(node))
    .flatMap((node): Change[] => {
      if (
        importPath === '@angular/common' &&
        isSymbolInClassMetadataFieldArray(node, metadataField, 'CommonModule')
      ) {
        // Special case: don't add children of CommonModule if it's already present.
        return [];
      }

      // Get all the children property assignment of object literals.
      const matchingProperties = getMetadataField(node, metadataField);

      if (!matchingProperties.length) {
        // We haven't found the field in the metadata declaration. Insert a new field.
        let position: number;
        let toInsert: string;
        if (!node.properties.length) {
          position = node.getEnd() - 1;
          toInsert = `\n  ${metadataField}: [\n${tags.indentBy(4)`${symbolName}`}\n  ]\n`;
        } else {
          const childNode = node.properties[node.properties.length - 1];
          position = childNode.getEnd();
          // Get the indentation of the last element, if any.
          const text = childNode.getFullText(source);
          const matches = text.match(/^(\r?\n)(\s*)/);
          if (matches) {
            toInsert =
              `,${matches[0]}${metadataField}: [${matches[1]}` +
              `${tags.indentBy(matches[2].length + 2)`${symbolName}`}${matches[0]}]`;
          } else {
            toInsert = `, ${metadataField}: [${symbolName}]`;
          }
        }
        if (importPath !== null && !insertedImport) {
          insertedImport = true;
          return [
            new InsertChange(filePath, position, toInsert),
            insertImport(
              source,
              filePath,
              // Remove function call syntax (e.g., SomeSymbol()) and property access (e.g., SomeSymbol.method) from symbolName for import.
              symbolName.replace(/(\.\w+|\(.*\))$/, ''),
              importPath,
            ),
          ];
        } else {
          return [new InsertChange(filePath, position, toInsert)];
        }
      }
      const assignment = matchingProperties[0];

      // If it's not an array, nothing we can do really.
      if (
        !ts.isPropertyAssignment(assignment) ||
        !ts.isArrayLiteralExpression(assignment.initializer)
      ) {
        return [];
      }

      let expression: ts.Expression | ts.ArrayLiteralExpression;
      const assignmentInit = assignment.initializer;
      const elements = assignmentInit.elements;

      if (elements.length) {
        const symbolsArray = elements.map(
          (node) => tags.oneLine`${node.getText()}`,
        );
        if (symbolsArray.includes(tags.oneLine`${symbolName}`)) {
          return [];
        }

        expression = elements[elements.length - 1];
      } else {
        expression = assignmentInit;
      }

      let toInsert: string;
      let position = expression.getEnd();
      if (ts.isArrayLiteralExpression(expression)) {
        // We found the field but it's empty. Insert it just before the `]`.
        position--;
        toInsert = `\n${tags.indentBy(4)`${symbolName}`}\n  `;
      } else {
        // Get the indentation of the last element, if any.
        const text = expression.getFullText(source);
        const matches = text.match(/^(\r?\n)(\s*)/);
        if (matches) {
          toInsert = `,${matches[1]}${tags.indentBy(matches[2].length)`${symbolName}`}`;
        } else {
          toInsert = `, ${symbolName}`;
        }
      }

      if (importPath !== null && !insertedImport) {
        insertedImport = true;
        return [
          new InsertChange(filePath, position, toInsert),
          insertImport(
            source,
            filePath,
            symbolName.replace(/[.(].*$/, ''),
            importPath,
          ),
        ];
      }

      return [new InsertChange(filePath, position, toInsert)];
    });
}

/**
 * Checks if a symbol is present in a class metadata field array.
 */
export function isSymbolInClassMetadataFieldArray(
  node: ts.ObjectLiteralExpression,
  metadataField: string,
  symbolName: string,
): boolean {
  const matchingProperties = getMetadataField(node, metadataField);

  if (!matchingProperties.length) {
    return false;
  }
  const assignment = matchingProperties[0];

  if (
    !ts.isPropertyAssignment(assignment) ||
    !ts.isArrayLiteralExpression(assignment.initializer)
  ) {
    return false;
  }

  const assignmentInit = assignment.initializer;
  const elements = assignmentInit.elements;

  const symbolsArray = elements.map((node) => tags.oneLine`${node.getText()}`);
  return symbolsArray.includes(tags.oneLine`${symbolName}`);
}
