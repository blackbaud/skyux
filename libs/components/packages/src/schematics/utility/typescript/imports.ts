/**
 * The contents of this file are heavily inspired by Angular's source code.
 * @see https://github.com/angular/angular/blob/14.3.x/packages/core/schematics/utils/typescript/imports.ts
 */
import * as ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';

export interface ImportUsage {
  node: ts.Node;
  importName: string;
}

export function getImports(
  sourceFile: ts.SourceFile,
  packages: Map<string, string[]>
): ts.ImportSpecifier[] {
  const importSpecifiers: ts.ImportSpecifier[] = [];

  for (const [packageName, classes] of packages.entries()) {
    for (const cc of classes) {
      const specifiers = getImportSpecifiers(sourceFile, packageName, cc);

      if (specifiers) {
        importSpecifiers.push(...specifiers);
      }
    }
  }

  return importSpecifiers;
}

export function getUsages(
  sourceFile: ts.SourceFile,
  importSpecifier: ts.ImportSpecifier
): ImportUsage[] {
  const usages: ImportUsage[] = [];

  const visitNode = (node: ts.Node) => {
    // Skip this node and all of its children; imports are a special case.
    if (ts.isImportSpecifier(node)) {
      return;
    }

    const importName = importSpecifier.getText(sourceFile);

    if (ts.isIdentifier(node) && node.escapedText === importName) {
      usages.push({ node, importName });
    }

    ts.forEachChild(node, visitNode);
  };

  ts.forEachChild(sourceFile, visitNode);

  return usages;
}

/**
 * Gets top-level import specifiers with a specific name that is imported from a particular module.
 * E.g. given a file that looks like:
 *
 * ```
 * import { Component, Directive } from '@angular/core';
 * import { Foo } from './foo';
 * ```
 *
 * Calling `getImportSpecifiers(sourceFile, '@angular/core', 'Directive')` will yield the nodes
 * referring to `Directive` in the top import.
 *
 * @param sourceFile File in which to look for imports.
 * @param moduleName Name of the import's module.
 * @param specifierName Original name of the specifier to look for. Aliases will be resolved to
 *    their original name.
 */
function getImportSpecifiers(
  sourceFile: ts.SourceFile,
  moduleName: string,
  specifierName: string
): ts.ImportSpecifier[] {
  const importSpecifiers: ts.ImportSpecifier[] = [];

  for (const node of sourceFile.statements) {
    if (
      ts.isImportDeclaration(node) &&
      ts.isStringLiteral(node.moduleSpecifier) &&
      node.moduleSpecifier.text === moduleName
    ) {
      const namedBindings =
        node.importClause && node.importClause.namedBindings;

      if (namedBindings && ts.isNamedImports(namedBindings)) {
        const match = findImportSpecifier(
          namedBindings.elements,
          specifierName
        );

        if (match) {
          importSpecifiers.push(match);
        }
      }
    }
  }

  return importSpecifiers;
}

/**
 * Finds an import specifier with a particular name.
 */
function findImportSpecifier(
  nodes: ts.NodeArray<ts.ImportSpecifier>,
  specifierName: string
): ts.ImportSpecifier | undefined {
  return nodes.find((element) => {
    const { name, propertyName } = element;

    return propertyName?.text === specifierName || name.text === specifierName;
  });
}
