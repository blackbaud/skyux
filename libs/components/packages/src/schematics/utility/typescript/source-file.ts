import * as ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';

export function createSourceFile(
  filePath: string,
  fileContents: string,
): ts.SourceFile {
  return ts.createSourceFile(
    filePath,
    fileContents,
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
}
