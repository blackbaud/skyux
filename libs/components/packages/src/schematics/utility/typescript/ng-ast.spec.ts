import { stripIndents } from '@angular-devkit/core/src/utils/literals';
import { HostTree } from '@angular-devkit/schematics';
import * as ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';
import { getDecoratorMetadata } from '@schematics/angular/utility/ast-utils';
import { Change, InsertChange } from '@schematics/angular/utility/change';

import {
  addSymbolToClassMetadata,
  getInlineTemplates,
  getTemplateUrls,
  isSymbolInClassMetadataFieldArray,
} from './ng-ast';

function getTsSource(path: string, content: string): ts.SourceFile {
  return ts.createSourceFile(path, content, ts.ScriptTarget.Latest, true);
}

function applyChanges(
  path: string,
  content: string,
  changes: Change[],
): string {
  const tree = new HostTree();
  tree.create(path, content);
  const exportRecorder = tree.beginUpdate(path);
  for (const change of changes) {
    if (change instanceof InsertChange) {
      exportRecorder.insertLeft(change.pos, change.toAdd);
    }
  }
  tree.commitUpdate(exportRecorder);

  return tree.readText(path);
}

describe('ng-ast', () => {
  describe('getInlineTemplates', () => {
    it('should get inline templates', () => {
      const sourceText = stripIndents`
      import { Component } from '@angular/core';

      @Component({
        selector: 'app-test',
        template: '<div>Hello</div>'
      })
      export class TestComponent {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      expect(getInlineTemplates(sourceFile)).toEqual([{ start: 90, end: 106 }]);
      expect(sourceText.slice(90, 106)).toBe('<div>Hello</div>');
    });

    it('should not get inline templates if it is not a string', () => {
      const sourceText = stripIndents`
      import { Component } from '@angular/core';

      @Component({
        selector: 'app-test',
        template: false,
      })
      export class TestComponent {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      expect(getInlineTemplates(sourceFile)).toEqual([]);
    });

    it('should not get inline templates if it is not an Angular component', () => {
      const sourceText = stripIndents`
      import { NgModule } from '@angular/core';

      @NgModule({})
      export class TestModule {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      expect(getInlineTemplates(sourceFile)).toEqual([]);
    });
  });

  describe('getTemplateUrls', () => {
    it('should get template URLs', () => {
      const sourceText = stripIndents`
      import { Component } from '@angular/core';

      @Component({
        selector: 'app-test',
        templateUrl: './test.component.html'
      })
      export class TestComponent {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      expect(getTemplateUrls(sourceFile)).toEqual(['./test.component.html']);
    });

    it('should not get template URL if it is not a string', () => {
      const sourceText = stripIndents`
      import { Component } from '@angular/core';

      @Component({
        selector: 'app-test',
        templateUrl: false,
      })
      export class TestComponent {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      expect(getTemplateUrls(sourceFile)).toEqual([]);
    });

    it('should not get template URL if it is not an Angular component', () => {
      const sourceText = stripIndents`
      import { NgModule } from '@angular/core';

      @NgModule({})
      export class TestModule {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      expect(getTemplateUrls(sourceFile)).toEqual([]);
    });
  });

  it('should getTestingModuleMetadata', () => {
    const sourceText = stripIndents`
      import { TestBed } from '@angular/core/testing';

      describe('My Test', () => {
        beforeEach(() => {
          TestBed.configureTestingModule({
            declarations: [AppComponent],
            imports: [CommonModule],
          });
        });
      });

      describe('My Other Test', () => {
        beforeEach(() => {
          TestBed.configureTestingModule({
            declarations: [OtherComponent],
          });
        });
      });
    `;
    const sourceFile = ts.createSourceFile(
      'test.ts',
      sourceText,
      ts.ScriptTarget.Latest,
      true,
      ts.ScriptKind.TS,
    );
    const { getTestingModuleMetadata } = require('./ng-ast');
    const metadata = getTestingModuleMetadata(sourceFile);
    expect(metadata.length).toBe(2);
    expect(metadata[0].getText()).toContain('declarations: [AppComponent]');
    expect(metadata[0].properties[1].getText()).toEqual(
      'imports: [CommonModule]',
    );
    expect(metadata[1].getText()).toContain('declarations: [OtherComponent]');
  });

  describe('addSymbolToClassMetadata', () => {
    let modulePath: string;
    let moduleContent: string;

    beforeEach(() => {
      modulePath = '/src/app/app.module.ts';
      moduleContent = `
      import { BrowserModule } from '@angular/platform-browser';
      import { NgModule } from '@angular/core';
      import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule
        ],
        providers: [],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;
    });

    it('should add metadata', () => {
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'HelloWorld',
      );
      expect(changes).not.toBeNull();

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toMatch(/imports: \[[^\]]+,\n(\s*) {2}HelloWorld\n\1\]/);
    });

    it('should add metadata (comma)', () => {
      const moduleContent = `
      import { BrowserModule } from '@angular/platform-browser';
      import { NgModule } from '@angular/core';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
        ],
        providers: [],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'HelloWorld',
      );
      expect(changes).not.toBeNull();

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toMatch(/imports: \[[^\]]+,\n(\s*) {2}HelloWorld,\n\1\]/);
    });

    it('should add metadata (missing)', () => {
      const moduleContent = `
      import { BrowserModule } from '@angular/platform-browser';
      import { NgModule } from '@angular/core';

      @NgModule({
        declarations: [
          AppComponent
        ],
        providers: [],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'HelloWorld',
      );
      expect(changes).not.toBeNull();

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toMatch(/imports: \[\n(\s*) {2}HelloWorld\n\1\]/);
    });

    it('should add metadata (empty)', () => {
      const moduleContent = `
      import { BrowserModule } from '@angular/platform-browser';
      import { NgModule } from '@angular/core';

      @NgModule({
        declarations: [
          AppComponent
        ],
        providers: [],
        imports: [],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'HelloWorld',
      );
      expect(changes).not.toBeNull();

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toMatch(/imports: \[\n(\s*) {2}HelloWorld\n\1\],\n/);
    });

    it('should add metadata (match spacing)', () => {
      const moduleContent = `
      import { NgModule } from '@angular/core';

      @NgModule({ declarations: [AppComponent] })
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'HelloWorld',
      );
      expect(changes).not.toBeNull();

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toContain(
        '({ declarations: [AppComponent], imports: [HelloWorld] })',
      );
    });

    it('should add metadata (no previous properties)', () => {
      const moduleContent = `
      import { NgModule } from '@angular/core';

      @NgModule({})
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'HelloWorld',
        './hello-world',
      );
      expect(changes).not.toBeNull();

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toMatch(/\{\n {2}imports: \[\n {4}HelloWorld\n {2}\]\n}/);
    });

    it('should not add metadata if the value is not an array expression', () => {
      const moduleContent = `
      import { NgModule } from '@angular/core';

      @NgModule({
        imports: null,
      })
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'HelloWorld',
      );
      expect(changes).toEqual([]);
    });

    it('should not add metadata if the value is already present', () => {
      const moduleContent = `
      import { NgModule } from '@angular/core';

      @NgModule({
        imports: [HelloWorld],
      })
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'HelloWorld',
      );
      expect(changes).toEqual([]);
    });

    it('should skip adding @angular/common services if CommonModule is already present', () => {
      const moduleContent = `
      import { BrowserModule } from '@angular/platform-browser';
      import { NgModule } from '@angular/core';
      import { CommonModule } from '@angular/common';

      @NgModule({
        declarations: [
          AppComponent
        ],
        providers: [],
        imports: [CommonModule],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'NgIf',
        '@angular/common',
      );
      expect(changes).toEqual([]);
    });

    it('should add CommonModule', () => {
      const moduleContent = `
      import { BrowserModule } from '@angular/platform-browser';
      import { NgModule } from '@angular/core';

      @NgModule({
        declarations: [
          AppComponent
        ],
        providers: [],
        imports: [],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'NgModule',
        modulePath,
        'imports',
        'CommonModule',
        '@angular/common',
      );
      expect(changes).not.toBeNull();
      expect(changes).not.toEqual([]);

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toContain('CommonModule');
      expect(output).toMatchSnapshot();
    });

    it('should add AsyncPipe', () => {
      const moduleContent = `
      import { Component } from '@angular/core';

      @Component({
        selector: 'app-test',
        template: '<div>Hello</div>',
      })
      export class AppTest { }
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'Component',
        modulePath,
        'imports',
        'AsyncPipe',
        '@angular/common',
      );
      expect(changes).not.toBeNull();
      expect(changes).not.toEqual([]);

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toContain('AsyncPipe');
      expect(output).toMatchSnapshot();
    });

    it('should add metadata to TestBed.configureTestingModule', () => {
      const moduleContent = stripIndents`
      import { TestBed } from '@angular/core/testing';

      describe('My Test', () => {
        beforeEach(() => {
          TestBed.configureTestingModule({
            declarations: [AppComponent],
            imports: [CommonModule],
          });
        });
      });

      describe('My Other Test', () => {
        beforeEach(() => {
          TestBed.configureTestingModule({
            declarations: [OtherComponent],
          });
        });
      });
    `;
      const source = getTsSource(modulePath, moduleContent);
      const changes = addSymbolToClassMetadata(
        source,
        'TestBed.configureTestingModule',
        modulePath,
        'imports',
        'HelloWorld',
        './hello-world',
      );
      expect(changes).not.toBeNull();
      expect(changes).not.toEqual([]);

      const output = applyChanges(modulePath, moduleContent, changes || []);
      expect(output).toContain('imports: [CommonModule, HelloWorld]');
      expect(output).toMatchSnapshot();
    });
  });

  describe('isSymbolInClassMetadataFieldArray', () => {
    it('should find import', () => {
      const sourceText = stripIndents`
      import { Component } from '@angular/core';
      import { AsyncPipe } from '@angular/common';

      @Component({
        selector: 'app-test',
        template: '<div>Hello</div>',
        imports: [AsyncPipe],
      })
      export class TestComponent {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      const metadata = getDecoratorMetadata(
        sourceFile,
        'Component',
        '@angular/core',
      );
      expect(metadata.length).toBe(1);
      expect(
        isSymbolInClassMetadataFieldArray(
          metadata[0] as any,
          'imports',
          'AsyncPipe',
        ),
      ).toBeTruthy();
    });

    it('should not find import when there are no imports', () => {
      const sourceText = stripIndents`
      import { Component } from '@angular/core';

      @Component({
        selector: 'app-test',
        template: '<div>Hello</div>',
      })
      export class TestComponent {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      const metadata = getDecoratorMetadata(
        sourceFile,
        'Component',
        '@angular/core',
      );
      expect(metadata.length).toBe(1);
      expect(
        isSymbolInClassMetadataFieldArray(
          metadata[0] as any,
          'imports',
          'AsyncPipe',
        ),
      ).toBeFalsy();
    });

    it('should not find import when imports is not an array expression', () => {
      const sourceText = stripIndents`
      import { Component } from '@angular/core';

      @Component({
        selector: 'app-test',
        template: '<div>Hello</div>',
        imports: null,
      })
      export class TestComponent {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      const metadata = getDecoratorMetadata(
        sourceFile,
        'Component',
        '@angular/core',
      );
      expect(metadata.length).toBe(1);
      expect(
        isSymbolInClassMetadataFieldArray(
          metadata[0] as any,
          'imports',
          'AsyncPipe',
        ),
      ).toBeFalsy();
    });

    it('should not find import in empty list', () => {
      const sourceText = stripIndents`
      import { Component } from '@angular/core';

      @Component({
        selector: 'app-test',
        template: '<div>Hello</div>',
        imports: [],
      })
      export class TestComponent {}
    `;
      const sourceFile = ts.createSourceFile(
        'test.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true,
        ts.ScriptKind.TS,
      );
      const metadata = getDecoratorMetadata(
        sourceFile,
        'Component',
        '@angular/core',
      );
      expect(metadata.length).toBe(1);
      expect(
        isSymbolInClassMetadataFieldArray(
          metadata[0] as any,
          'imports',
          'AsyncPipe',
        ),
      ).toBeFalsy();
    });
  });
});
