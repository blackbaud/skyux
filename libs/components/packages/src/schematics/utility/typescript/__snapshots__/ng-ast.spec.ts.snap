// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ng-ast addSymbolToClassMetadata should add AsyncPipe 1`] = `
"
      import { Component } from '@angular/core';
import { AsyncPipe } from '@angular/common';

      @Component({
        selector: 'app-test',
        template: '<div>Hello</div>',
        imports: [
          AsyncPipe
        ],
      })
      export class AppTest { }
    "
`;

exports[`ng-ast addSymbolToClassMetadata should add CommonModule 1`] = `
"
      import { BrowserModule } from '@angular/platform-browser';
      import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

      @NgModule({
        declarations: [
          AppComponent
        ],
        providers: [],
        imports: [
    CommonModule
  ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    "
`;

exports[`ng-ast addSymbolToClassMetadata should add metadata to TestBed.configureTestingModule 1`] = `
"import { TestBed } from '@angular/core/testing';
import { HelloWorld } from './hello-world';

describe('My Test', () => {
beforeEach(() => {
TestBed.configureTestingModule({
declarations: [AppComponent],
imports: [CommonModule, HelloWorld],
});
});
});

describe('My Other Test', () => {
beforeEach(() => {
TestBed.configureTestingModule({
declarations: [OtherComponent],
imports: [
  HelloWorld
],
});
});
});"
`;
